libdnf/swdb/item_comps_environment.cpp

```diff
@@ -0,0 +1,328 @@
+/*
+ * Copyright (C) 2017-2018 Red Hat, Inc.
+ *
+ * Licensed under the GNU Lesser General Public License Version 2.1
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
+ */
+
+#include <algorithm>
+
+#include "item_comps_environment.hpp"
+
+CompsEnvironmentItem::CompsEnvironmentItem(std::shared_ptr< SQLite3 > conn)
+  : Item{conn}
+{
+}
+
+CompsEnvironmentItem::CompsEnvironmentItem(std::shared_ptr< SQLite3 > conn, int64_t pk)
+  : Item{conn}
+{
+    dbSelect(pk);
+}
+
+void
+CompsEnvironmentItem::save()
+{
+    if (getId() == 0) {
+        dbInsert();
+    } else {
+        // dbUpdate();
+    }
+    for (auto i : getGroups()) {
+        i->save();
+    }
+}
+
+void
+CompsEnvironmentItem::dbSelect(int64_t pk)
+{
+    const char *sql = R"**(
+        SELECT
+            environmentid,
+            name,
+            translated_name,
+            pkg_types
+        FROM
+            comps_environment
+        WHERE
+            item_id = ?
+    )**";
+    SQLite3::Query query(*conn.get(), sql);
+    query.bindv(pk);
+    query.step();
+
+    setId(pk);
+    setEnvironmentId(query.get< std::string >("environmentid"));
+    setName(query.get< std::string >("name"));
+    setTranslatedName(query.get< std::string >("translated_name"));
+    setPackageTypes(static_cast< CompsPackageType >(query.get< int >("pkg_types")));
+}
+
+void
+CompsEnvironmentItem::dbInsert()
+{
+    // populates this->id
+    Item::save();
+
+    const char *sql = R"**(
+        INSERT INTO
+            comps_environment (
+                item_id,
+                environmentid,
+                name,
+                translated_name,
+                pkg_types
+            )
+        VALUES
+            (?, ?, ?, ?, ?)
+    )**";
+    SQLite3::Statement query(*conn.get(), sql);
+    query.bindv(getId(),
+                getEnvironmentId(),
+                getName(),
+                getTranslatedName(),
+                static_cast< int >(getPackageTypes()));
+    query.step();
+}
+
+static std::shared_ptr< TransactionItem >
+compsEnvironmentTransactionItemFromQuery(std::shared_ptr< SQLite3 > conn, SQLite3::Query &query)
+{
+    auto trans_item = std::make_shared< TransactionItem >(conn);
```

Comment:
@dmach There is no `std::shared_ptr< SQLite3 >` constructor for `TransactionItem` - not sure why compiler allows this.